### PR TITLE
Add error autocompletion

### LINKED ISSUE #9107: Enable auto-completion of error fields
[Link to the design document](https://www.notion.so/tarantool/Error-subsystem-improvements-90faa0a4714b4143abaf8bed2c10b2fc?pvs=4#cd10a63b3a184944ba06b1315f95ba91)

To make an error object easier to use, we’ll implement the `__autocomplete` method. The method will enable auto-completion of the error object fields in the console. We’ll enable auto-completion for both built-in and payload fields.

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
This patch set adds error autocompletion.

Closes #9107

### CODE REVIEW

**Review by Gumix — COMMENTED** [review]
Please add payload fields of the cause to the autocompletion:
```diff
diff --git a/src/lua/error.lua b/src/lua/error.lua
index 3b4b7b703..9ebb07587 100644
--- a/src/lua/error.lua
+++ b/src/lua/error.lua
@@ -174,7 +174,16 @@ local error_methods = {
 }

 local function error_autocomplete(err)
-    local err_unpacked = err:unpack()
+    local err_unpacked = {}
+    local cur_err = err
+    while cur_err ~= nil do
+        for key, val in pairs(cur_err:unpack()) do
+            if err_unpacked[key] == nil then
+                err_unpacked[key] = val
+            end
+        end
+        cur_err = cur_err.prev
+    end
     for key, method in pairs(error_methods) do
         if key:match('__.+') == nil then
             err_unpacked[key] = method
```

**Comment by CuriousGeorgiy** [comment]
@Gumix Done.

**Review by ochaplashkin — APPROVED** [review]

File: src/lua/error.lua [hunk 1]
```diff
@@ -173,6 +173,29 @@ local error_methods = {
...
+
+    for key, method in pairs(error_methods) do
+        if key:match('__.+') == nil then
+            err_unpacked[key] = method
+        end
+    end
+    return err_unpacked
+end
+
+error_methods["__autocomplete"] = error_autocomplete
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/lua/error.lua:197]: I access the `error_methods` table from `error_autocomplete`, so I need to declare the former earlier than the latter.

File: test/box-luatest/error_subsystem_improvements_test.lua [hunk 2]
```diff
@@ -138,3 +139,42 @@ g.test_payload_inheritance = function()
...
+                              'err.prev',
+                              'err.message',
+                              'err.match(',
+                              'err.set_prev(',
+                              'err.trace',
+                              'err.type',
+                              'err.unpack(',
+                              'err.code',
+                              'err.bar',
+                              })
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply test/box-luatest/error_subsystem_improvements_test.lua]: I was looking at this test as a reference:
https://github.com/tarantool/tarantool/blob/ee08747291788868ddd5fe5f0571b48c2501b89b/test/box-luatest/gh_6305_net_box_autocomplete_test.lua#L29-L50

Moved the `})` to the beginning of the line and aligned all table values with it.

File: src/lua/error.lua [hunk 3]
```diff
@@ -173,6 +173,29 @@ local error_methods = {
...
+        for key, val in pairs(cur_err:unpack()) do
+            if err_unpacked[key] == nil then
+                err_unpacked[key] = val
+            end
+        end
+        cur_err = cur_err.prev
+    end
+
+    for key, method in pairs(error_methods) do
+        if key:match('__.+') == nil then
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/lua/error.lua]: Sure, I didn't know about `startswith`, thanks!

File: src/box/lua/console.c [hunk 4]
```diff
@@ -1105,9 +1129,9 @@ lua_rl_getcompletion(lua_State *L)
 		lua_pop(L, 1);
 		return 0;
 	}
-	/* use __autocomplete metamethod if it's present */
-	lua_pushstring(L, "__autocomplete");
-	lua_rawget(L, -2);
+	/* use __autocomplete method if it's present */
+	int top = lua_gettop(L);
+	lua_rl_get_autocomplete_metamethod(L, top - 1, top);
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/lua/console.c]: Fixed.

File: src/lua/error.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (nshy) [reply src/lua/error.lua:210]: We can declare `error_methods` (`local error_methods`) before `error_autocomplete` instead.

**Review by nshy — APPROVED** [review]

**Comment by sergepetrenko** [comment]
@andreyaksenov, PTAL on behalf of @tarantool/doc

**Review by andreyaksenov — APPROVED** [review]

**Pull request merged**